docs: #866 Update Intro to Redpanda (#987)
redpanda-data/documentation: docs/platform/introduction/intro-to-events.mdx

---
title: Introduction to Redpanda
---

<head>
    <meta name="title" content="Intro events and streaming | Redpanda Docs"/>
    <meta name="description" content="How events improve your app: An intro to events and streaming"/>
</head>

Distributed systems often require data and system updates to happen as quickly as possible. In software architecture, these updates can be handled with either messages or events. 

- With messages, updates are sent directly from one component to another to trigger an action.

> - With events, updates indicates that an action occurred at a specific time. They are not directed to a specific recipient.

Reviewer: ```suggestion
- With events, updates indicate that an action occurred at a specific time. They are not directed to a specific recipient.
```

An event is simply a record of something changing state. For example, the event of a credit card transaction includes the product purchased, the payment, the delivery, and the time of the purchase. The event occurred in the purchasing component, but it also impacts the inventory, the payment processing, and the shipping components.

> In event-driven architecture, all actions are defined and packaged as events to precisely identify individual actions and how they're processed throughout the system. Instead of processing updates in consecutive order, event-driven architecture lets components process events at their own pace. This helps developers build fast and scalable systems.

Reviewer: My understanding is that within Kafka, events are actions or operations (for example, authentication events, or authorization events), and that Kafka publishes streams of these events. What I'm not sure of, though, is whether events as you have categorized them here (purchase, payment, delivery, inventory, etc) map to events as defined in Kafka. I definitely think it is worthwhile to have a PM weigh in on the analogy here to see if it is apt for our audience. I could be out in left field on this one, but that is the thought I had after reading this.

## What is Redpanda?

Redpanda is an event streaming platform: it provides the infrastructure for streaming real-time data. 

> Producers are client applications that write data to Redpanda in the form of events. Redpanda reads the events in sequence and organizes them into topics, representing a replayable log of changes in the system. 

Reviewer: ```suggestion
Producers are client applications that write data to Redpanda in the form of events. Redpanda reads the events in sequence and organizes them into topics, which represent a replayable log of changes in the system. 
```

> Consumers are client applications that then read these events from Redpanda topics. Consumers can store, process, or react to the events. 

Reviewer: ```suggestion
Consumers are client applications that read these events from Redpanda topics. Consumers can store, process, or react to the events. 
```

Redpanda decouples producers from consumers to allow for asynchronous event processing, event tracking, event manipulation, and event archiving. Producers and consumers interact with Redpanda using the Apache Kafka® API.  

> | Event-driven updates | Message-driven updates |

Reviewer: Boy, this would be a fantastic place for a diagram that depicts this difference. Have you thought about that? We can request through Ben...let me know if you want to do that. I just think a visual representation of this that a reader could scan would reinforce/clarify the difference.

Author: @Feediver1: I agree! This page could use an image to visualize benefits of Redpanda.

Reviewer 2: @micheleRP I think the table should make it clear to the reader that Redpanda only supports events. Redpanda doesn't support sending messages directly from a producer to a consumer.

| -------- | ------ |
| Events are sent from the producer to an event processing system (Redpanda) that acknowledges receipt and guarantees delivery to multiple consumers. The system tracks communication between producers and consumers if there's a broken connection. | Messages are sent directly from the producer to each consumer. The producer must wait for acknowledgement that the consumer received the message before it can continue with its processes. If there's a break in the connection between a producer and a recipient, the producer doesn’t know if the message was processed or if it needs to resend it. |

Event streaming lets you extract value out of each event by analyzing, mining, or transforming it for insights. You can:

- Take one event and consume it in multiple ways.
- Replay events from the past and route them to new processes in your application.
- Run transformations on the data in real-time or historically.
- Integrate with other event processing systems that use the Kafka API.

## Redpanda differentiators

> Redpanda is less complex and less costly than any other mission-critical event streaming platform. It's fast, it's easy, and it keeps your data safe. 

Reviewer: Please run this by PM. We have to be 100% sure this claim is absolutely true.

Author: @mattschumpert Would you please review this? I know we say it, and post it... but this OK to put in our docs?

Reviewer 3: we might want to add the word 'commercial' to event streaming platform.  One can certainly make an argument that Apache Kafka (free software) can be used for mission-critical event streaming.

> - Redpanda is designed for maximum performance on any data streaming workload. It can scale up to use all available resources on a single machine and scale out to distribute performance across multiple machines. This enables previously-unimaginable use cases that require high throughput, low latency, and a minimal hardware footprint. 

Reviewer: How about setting each of the first statements off from the rest of the paragraph for emphasis and clarity wrt to our key claims?
- Redpanda is designed for maximum performance on any data streaming workload. 
   Redpanda can scale up to use all available resources on a single machine and scale out to distribute 
   performance across multiple machines. This enables previously-unimaginable use cases that require high 
   throughput, low latency, and a minimal hardware footprint.

> - Redpanda is packaged as a single binary: it doesn't rely on any external systems. It's fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations built on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or in the cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. Additionally, Tiered Storage lets you archive log segments to cloud storage in near real time, providing infinite data retention and topic recovery. 

Reviewer: ```suggestion
- Redpanda is packaged as a single binary: it doesn't rely on any external systems. 
   Redpanda is fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations built 
   on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or in the 
   cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. Additionally, Tiered 
   Storage lets you archive log segments to cloud storage in near real time, providing infinite data retention and 
   topic recovery. 
```

> - Redpanda is packaged as a single binary: it doesn't rely on any external systems. It's fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations built on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or in the cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. Additionally, Tiered Storage lets you archive log segments to cloud storage in near real time, providing infinite data retention and topic recovery. 

Reviewer: - Redpanda is packaged as a single binary: it doesn't rely on any external systems. 
   Redpanda is fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations 
   built on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or 
   in the cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. 
   Additionally, Tiered Storage lets you archive log segments to cloud storage in near real time, providing 
   infinite data retention and topic recovery.

Reviewer: ignore formatting grimness--you get the gist of it...

> - Redpanda uses the [Raft consensus algorithm](https://raft.github.io/) to write data to log files and replicate that data across multiple servers. Raft facilitates communication between the nodes in a Redpanda cluster to make sure that they agree on changes and remain in sync, even if a minority of them are in a failure state. This allows Redpanda to tolerate partial environmental failures and deliver predictable performance, even at high loads.

Reviewer: - Redpanda uses the [Raft consensus algorithm](https://raft.github.io/) to write data to log files and replicate that data across multiple servers. 
Raft facilitates communication between the nodes in a Redpanda cluster to make sure that they agree on changes and remain in sync, even if a minority of them are in a failure state. This allows Redpanda to tolerate partial environmental failures and deliver predictable performance, even at high loads.

In event-driven architecture, all actions are defined and packaged as events to precisely identify individual actions and how they're processed throughout the system. Instead of processing updates in consecutive order, event-driven architecture lets components process events at their own pace. This helps developers build fast and scalable systems.

## What is Redpanda?

Redpanda is an event streaming platform: it provides the infrastructure for streaming real-time data. 

Producers are client applications that write data to Redpanda in the form of events. Redpanda reads the events in sequence and organizes them into topics, which represent a replayable log of changes in the system. 

Consumers are client applications that read these events from Redpanda topics. Consumers can store, process, or react to the events. 

Redpanda decouples producers from consumers to allow for asynchronous event processing, event tracking, event manipulation, and event archiving. Producers and consumers interact with Redpanda using the Apache Kafka® API.  

| Event-driven updates | Message-driven updates |
| -------- | ------ |

> | Events are sent from the producer to an event processing system (Redpanda) that acknowledges receipt and guarantees delivery to multiple consumers. The system tracks communication between producers and consumers if there's a broken connection. | Messages are sent directly from the producer to each consumer. The producer must wait for acknowledgement that the consumer received the message before it can continue with its processes. If there's a break in the connection between a producer and a recipient, the producer doesn’t know if the message was processed or if it needs to resend it. |

Reviewer 2: > _Events are sent from the producer to an event processing system (Redpanda) that acknowledges receipt and guarantees delivery to multiple consumers. The system tracks communication between producers and consumers if there's a broken connection._

From a producer perspective, Redpanda will acknowledge receipt of a write and this provides inherent guarantees that the write is durable within the system. A producer is not aware of any consumers, they are completely decoupled. Redpanda does guarantee delivery to the subscribed consumers, but I'm not sure what you mean by the system tracks communication between producers and consumers if there is a broken connection.

Author: Thanks, @jrkinley. I simplified this table. We can remove it altogether if it's more confusing than helpful.

Event streaming lets you extract value out of each event by analyzing, mining, or transforming it for insights. You can:

- Take one event and consume it in multiple ways.
- Replay events from the past and route them to new processes in your application.
- Run transformations on the data in real-time or historically.
- Integrate with other event processing systems that use the Kafka API.

## Redpanda differentiators

> Redpanda is less complex and less costly than any other commericial mission-critical event streaming platform. It's fast, it's easy, and it keeps your data safe. 

Reviewer 2: I think we should remove the word _commercial_ here otherwise we're ignoring Apache Kafka.

Author: @jrkinley: Matt had suggested that we add "commercial" to this line. See https://github.com/redpanda-data/documentation/pull/987#discussion_r1065249663.

  
- Redpanda is designed for maximum performance on any data streaming workload. 
  
  It can scale up to use all available resources on a single machine and scale out to distribute performance across multiple nodes. Built on C++, Redpanda delivers greater throughput and up to 10x lower p99 latencies than other platforms. This enables previously-unimaginable use cases that require high throughput, low latency, and a minimal hardware footprint. 
  
- Redpanda is packaged as a single binary: it doesn't rely on any external systems. 
  

>   It's fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations built on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or in the cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. Additionally, Tiered Storage lets you archive log segments to cloud storage in near real time, providing infinite data retention and topic recovery. 

Reviewer 2: > _It's fully compatible with the Kafka API_

It's compatible with the Kafka API

Redpanda is less complex and less costly than any other mission-critical event streaming platform. It's fast, it's easy, and it keeps your data safe. 
  
- Redpanda is designed for maximum performance on any data streaming workload. 
  
  It can scale up to use all available resources on a single machine and scale out to distribute performance across multiple machines. Built on C++, Redpanda delivers greater throughput and up to 10x lower p99 latencies than other platforms. This enables previously-unimaginable use cases that require high throughput, low latency, and a minimal hardware footprint. 
  
- Redpanda is packaged as a single binary: it doesn't rely on any external systems. 
  
  It's fully compatible with the Kafka API, so it works with the ecosystem of tools and integrations built on Kafka. Redpanda can be deployed on bare metal, containers, or virtual machines in a data center or in the cloud. And Redpanda Console makes it easy to set up, manage, and monitor your clusters. Additionally, Tiered Storage lets you archive log segments to cloud storage in near real time, providing infinite data retention and topic recovery. 
  

> - Redpanda uses the [Raft consensus algorithm](https://raft.github.io/) to write data to log files and replicate that data across multiple servers. 

Reviewer 4: Is it more correct to state: "Redpanda uses the Raft consensus algorithm to _coordinate_ writing data to log files ..." ?

Author: Thanks, @hcoyote! "How's this: Redpanda uses the [Raft consensus algorithm](https://raft.github.io/) throughout the platform to coordinate writing data to log files and replicating that data across multiple servers."

Reviewer 4: +1
